### DISCUSSION TITLE
Learning method of eth_mnist.py and results

### DISCUSSION THREAD

**Discussion by rafaelblevin821** [BindsNET/bindsnet#574] [General]
Hi, so I know the eth_mnist.py network uses the Diehl and Cook 2015 method with STDP, but how does it actually sort the digits into classes (such as K-means or K nearest neighbour clustering). Is there anywhere we can view the results and which images where incorrectly identified?

Also, is there any way the network accuracy can be saved so that it can start from, say, 85% accuracy on the next time it is loaded instead of starting from zero again?

Any help would be appreciated.

**Comment by Hananel-Hazan** [comment]
For more information on the training and testing method, please read [Unsupervised learning of digit recognition using spike-timing-dependent plasticity](https://www.frontiersin.org/article/10.3389/fncom.2015.00099) and for more details [Lattice map spiking neural networks (LM-SNNs) for clustering and classifying image data](https://link.springer.com/article/10.1007/s10472-019-09665-3) or [arXiv](https://arxiv.org/pdf/1906.11826)

To save the trained network use the operator `.save` ([docs](https://bindsnet-docs.readthedocs.io/bindsnet.network.html?highlight=save#bindsnet.network.Network.save)) and for loading the network use the operator `.load` ([docs](https://bindsnet-docs.readthedocs.io/bindsnet.network.html?highlight=save#bindsnet.network.load)). 

Because `eth_mnist.py` is trained in an unsupervised manner, you dont need anything else. When feeding new image to the loaded the network, you need to evaluate the spikes to determine which one of the neurons have more spikes and remap the neurons. To save time you can also save the `all_activity_pred`,  and `proportion_pred ` that contain that information. 

To identify the misclassified image compare the label of the data in `batch["label"]` verses the actual predicted label from (`all_activity_pred` or `proportion_pred`).

**Comment by rafaelblevin821** [comment]
Hi @Hananel-Hazan , thanks for the reply, great info.

`To identify the misclassified image compare the label of the data in batch["label"] verses the actual predicted label from (all_activity_pred or proportion_pred).`

Could you please let me know where I would find `all_activity_pred` and `proportion_pred`

**Comment by Hananel-Hazan** [comment]
`all_activity_pred` [here](https://github.com/BindsNET/bindsnet/blob/61fda794534bf3c85307f8f71a6124ff5c7c4b52/examples/mnist/eth_mnist.py#L190)
`proportion_pred` [here](https://github.com/BindsNET/bindsnet/blob/61fda794534bf3c85307f8f71a6124ff5c7c4b52/examples/mnist/eth_mnist.py#L193)

**Reply by rafaelblevin821** [reply]
> `all_activity_pred` [here](https://github.com/BindsNET/bindsnet/blob/61fda794534bf3c85307f8f71a6124ff5c7c4b52/examples/mnist/eth_mnist.py#L190) `proportion_pred` [here](https://github.com/BindsNET/bindsnet/blob/61fda794534bf3c85307f8f71a6124ff5c7c4b52/examples/mnist/eth_mnist.py#L193)

Hi @Hananel-Hazan, thanks again. Is there any way of saving these predictions to a text file?

**Reply by Hananel-Hazan** [reply]
yes, you can save it to text by converting the PyTorch no Numpy variable and then save it to text like this `np.savetxt("YOUR_FILE.csv", all_activity_pred.cpu().numpy(), delimiter=",")`

**Reply by rafaelblevin821** [reply]
> yes, you can save it to text by converting the PyTorch no Numpy variable and then save it to text like this `np.savetxt("YOUR_FILE.csv", all_activity_pred.cpu().numpy(), delimiter=",")`

Hi @Hananel-Hazan , thanks for your help again.
